### PR TITLE
Add data source of obs bucket object

### PR DESCRIPTION
**Pull request by smilemusic**
make testacc TEST=./huaweicloud/ TESTARGS='-run TestAccDataSourceAWSObsBucketObject_allParams'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/ -v -run TestAccDataSourceAWSObsBucketObject_allParams -timeout 360m
=== RUN   TestAccDataSourceAWSObsBucketObject_allParams
--- PASS: TestAccDataSourceAWSObsBucketObject_allParams (17.63s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       17.662s

 make testacc TEST=./huaweicloud/ TESTARGS='-run TestAccDataSourceObsBucketObject_source'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/ -v -run TestAccDataSourceObsBucketObject_source -timeout 360m
=== RUN   TestAccDataSourceObsBucketObject_source
--- PASS: TestAccDataSourceObsBucketObject_source (17.18s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       17.761s

### CODE REVIEW

File: huaweicloud/provider.go [hunk 1]
```diff
@@ -244,6 +244,7 @@ func Provider() terraform.ResourceProvider {
 			"huaweicloud_identity_role_v3":          dataSourceIdentityRoleV3(),
 			"huaweicloud_cdm_flavors_v1":            dataSourceCdmFlavorV1(),
 			"huaweicloud_dis_partition_v2":          dataSourceDisPartitionV2(),
+			"huaweicloud_obs_bucket_object":         dataSourceObsBucketObject(),
```
Comment (niuzhenguo) [inline huaweicloud/provider.go]: Please move this to L202.

**Review by niuzhenguo — COMMENTED** [review]
a small suggestion inline.

File: huaweicloud/provider.go (hunk 1 above)
Comment (smilemusic) [reply huaweicloud/provider.go]: > Please move this to L202.

ok

File: huaweicloud/data_source_huaweicloud_obs_bucket_object.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	log.Printf("[DEBUG] Data Source Reading OBS Bucket Object %s: %#v", key, object)
+
+	class := string(object.StorageClass)
+	if class == "" {
+		d.Set("storage_class", "STANDARD")
+	} else {
+		d.Set("storage_class", normalizeStorageClass(class))
+	}
+	d.SetId(uniqueId)
```
Comment (ShiChangkuo) [inline huaweicloud/data_source_huaweicloud_obs_bucket_object.go]: To be compatible with [obs object resource](https://www.terraform.io/docs/providers/huaweicloud/r/obs_bucket_object.html), setting id with the key.

File: huaweicloud/data_source_huaweicloud_obs_bucket_object.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+func dataSourceObsBucketObject() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceObsBucketObjectRead,
+
+		Schema: map[string]*schema.Schema{
+			"body": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
```
Comment (ShiChangkuo) [inline huaweicloud/data_source_huaweicloud_obs_bucket_object.go]: the recommended order in Schema: Required --> Optional --> Computed

File: huaweicloud/data_source_huaweicloud_obs_bucket_object.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	log.Printf("[DEBUG] Data Source Reading OBS Bucket Object : %#v", resp)
+
+	d.Set("content_type", resp.ContentType)
+	if resp.Body != nil {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return getObsError("Error get body of OBS bucket object", bucket, err)
+		}
+		d.Set("body", string(body))
+	}
```
Comment (ShiChangkuo) [inline huaweicloud/data_source_huaweicloud_obs_bucket_object.go]: The content of an object maybe very large, please do some test on a huge object, thanks

File: huaweicloud/data_source_huaweicloud_obs_bucket_object_test.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+	"os"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	"github.com/huaweicloud/golangsdk/openstack/obs"
+)
+
+func TestAccHuaweiCloudObsBucketObjectDataSource_content(t *testing.T) {
```
Comment (ShiChangkuo) [inline huaweicloud/data_source_huaweicloud_obs_bucket_object_test.go:15]: seems the `TestAccHuaweiCloudObsBucketObjectDataSource_content` and `TestAccHuaweiCloudObsBucketObjectDataSource_source` are the same testing case for data source.

File: huaweicloud/data_source_huaweicloud_obs_bucket_object_test.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+					testAccCheckObsBucketObjectExists("huaweicloud_obs_bucket_object.object"),
+				),
+			},
+			{
+				Config: dataSourceConf,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsObsObjectDataSourceExists("data.huaweicloud_obs_bucket_object.obj"),
+					resource.TestCheckResourceAttr("data.huaweicloud_obs_bucket_object.obj", "content_type", "binary/octet-stream"),
+					resource.TestCheckResourceAttr("data.huaweicloud_obs_bucket_object.obj", "storage_class", "STANDARD"),
+					resource.TestCheckResourceAttr("data.huaweicloud_obs_bucket_object.obj", "acl", ""),
```
Comment (ShiChangkuo) [inline huaweicloud/data_source_huaweicloud_obs_bucket_object_test.go]: why we got an empty `acl`?

File: huaweicloud/data_source_huaweicloud_obs_bucket_object.go (hunk 2 above)
Comment (smilemusic) [reply huaweicloud/data_source_huaweicloud_obs_bucket_object.go]: fixed

File: huaweicloud/data_source_huaweicloud_obs_bucket_object.go (hunk 3 above)
Comment (smilemusic) [reply huaweicloud/data_source_huaweicloud_obs_bucket_object.go]: fixed

File: huaweicloud/data_source_huaweicloud_obs_bucket_object_test.go (hunk 6 above)
Comment (smilemusic) [reply huaweicloud/data_source_huaweicloud_obs_bucket_object_test.go]: fixed

**Pull request merged**